File: JavaAlgorithm/src/main/java/algo/tzashinorpu/GSSY/DoublePoint/longestOnes_1004.java
package algo.tzashinorpu.GSSY.DoublePoint;

public class longestOnes_1004 {
    public int longestOnes(int[] nums, int k) {
        /*动态维护一个左右区间 [j, i] 和维护窗口内和 tot。
        右端点一直右移，左端点在窗口不满足「len - tol <= k」的时候进行右移*/
        int res = 0;
        for (int i = 0, j = 0, tot = 0; j < nums.length; j++) {
//            if (nums[j] == 1) tot++;
            tot += nums[j];
            while (j - i + 1 - tot > k) {
                /*i++;
                if (nums[i] == 1) tot--;*/
                tot -= nums[i++];
            }
            res = Math.max(res, j - i + 1);
        }
        return res;
    }
}

File: Clion/zero/1400_1600/longestOnes_1004.cpp
//
// Created by TzashiNorpu on 2023/6/12.
//
#include "../header/header.h"

using namespace zero;

int Solution::longestOnes(vector<int> &nums, int k) {
  int i = 0, j = 0;
  int n = nums.size();
  for (; j < n; j++) {
    if (nums[j] == 0) k--;
    if (k < 0 && nums[i++] == 0)k++;
  }
  return j - i;
}